Handle invalid `losetup' options

clickyotomy opened the pull request:
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
The non `util-linux` versions of `losetup` don't seem to have options like `-j` (`--associated`) which lists the loop devices associated with the file, or `--show` which  displays the name of the assigned loop device for a file.

For instance, when `-j` is used, `GetLoopDevice()` fails with:

```ShellSession
    $ losetup -j /path/to/file
    losetup: unrecognized option: j
    BusyBox v1.32.1 () multi-call binary.
```

Add a fallback option to lookup the device from `sysfs` in cases where `losetup` fails for an invalid option. This can be done by reading the backing file from `/sys/block/loop*/loop/backing_file` for each of the devices listed there.

#### Which issue(s) this PR fixes:
Fixes #83265

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs

```

/sig storage

msau42 commented:
/ok-to-test

msau42 commented:
/assign @jsafrane

clickyotomy commented:
/retest

clickyotomy commented:
/test pull-kubernetes-e2e-kind

(It passed earlier and failed after the last `retest`.)

clickyotomy commented:
Gentle ping @jsafrane, can you PTAL?

jsafrane commented on pkg/volume/storageos/storageos_util.go:
```diff
@@ -384,11 +389,22 @@ func getLoopDevice(path string, exec utilexec.Interface) (string, error) {
 	args := []string{"-f", "-P", "--show", path}
 	out, err := exec.Command(losetupPath, args...).CombinedOutput()
+	if err != nil {
+		if !losetupInvalidOpt(out) {
+			klog.V(2).Infof("Failed device create command for path %s: %v", path, err)
+			return "", err
+		}
```
Would it make more sense to just print  stdout/stderr returned by CombinedOutput and not try to analyze what it means?

clickyotomy replied:
Are you referring to printing `CombinedOutput` in `Infof` (which is a good idea) or checking the output? If it's the latter, we need it for checking if we have to lookup from `sysfs`.

jsafrane replied:
Just print whole `CombinedOutput` on error and do not parse it.

clickyotomy replied:
@jsafrane - The reason for parsing the output for an invalid argument is to conditionally call `getLoopDeviceFromSysfs`. If command failed for any other reason (like `EBUSY`, or something similar), shouldn't we just return the failure?

BTW, I have updated the function (in 3f15e65) to print the output in the logs.

jsafrane replied:
> The reason for parsing the output for an invalid argument is to conditionally call getLoopDeviceFromSysfs. If command failed for any other reason (like EBUSY, or something similar), shouldn't we just return the failure?

I don't expect `losetup -j` to return `EBUSY`. As you wrote below, it probably does read the same sysfs as your `getLoopDeviceFromSysfs`.

Problem with parsing the error message of `losetup` is that it may change unexpectedly in the future and it depends on current locale. We do parse error messages elsewhere in Kubernetes, but it's not a good idea to spread it more.

clickyotomy replied:
You're right, it would not return `EBUSY` for `-j`, for other options like `-f` it is a possibility. However, I do see that it becomes difficult to maintain of the the output is changed in the future.

I will remove `losetupInvalidOpt` and retry if it fails the first time. There's a good chance that if it failed for reasons other than an invalid option, it might fail again.

jsafrane commented on pkg/volume/storageos/storageos_util.go:
```diff
@@ -375,6 +376,10 @@ func getLoopDevice(path string, exec utilexec.Interface) (string, error) {
 	args := []string{"-j", path}
 	out, err := exec.Command(losetupPath, args...).CombinedOutput()
 	if err != nil {
+		if losetupInvalidOpt(out) {
+			return getLoopDeviceFromSysfs(path)
+		}
```
Is getLoopDeviceFromSysfs supported everywhere? Would it be better to use instead `losetup` in all cases?

clickyotomy replied:
@jsafrane - I think it can be substituted for just `-j` because `losetup` does something similar. In other cases like `-f --show`, for instance we'd have to call this function depending on the output.

(Also, thanks for taking the time to review this PR.)

clickyotomy commented:
/retest

clickyotomy committed:
Handle invalid "losetup" options

The non "util-linux" versions of "losetup" don't seem to have
options like "-j" ("--associated") which lists the loop devices
associated with the file, or "--show" which  displays the name
of the assigned loop device for a file.

For instance, when "-j" is used, "GetLoopDevice()" fails with:

    $ losetup -j /path/to/file
    losetup: unrecognized option: j
    BusyBox v1.32.1 () multi-call binary.

Add a fallback option to lookup the device from "sysfs" in cases
where "losetup" fails for an invalid option. This can be done by
reading the backing file from "/sys/block/loop*/loop/backing_file"
for each of the devices listed there.

Signed-off-by: Srinidhi Kaushik <shrinidhi.kaushik@gmail.com>

clickyotomy commented:
Summary of changes since the last commit:
 - Removed `losetupInvalidOpt` to avoid parsing error messages.
 - Removed `parseLosetupOutputForDevice` in favor of `getLoopDeviceFromSysfs`. Since the problematic options are `-j`, `--show` (for lookup and showing the device name after attach), we can lookup from `sysfs` in both cases directly, and avoid parsing the output as well.
 
/cc @jsafrane

clickyotomy commented:
/retest

jsafrane commented:
/test pull-kubernetes-node-e2e-containerd
/lgtm
/approve
Sorry about the delay!

clickyotomy commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd
